### PR TITLE
Keep a list of model instances in the JPA map session.

### LINKED ISSUE #12384: JPA map store uses bulk delete, but keeps entities in the persistence context
### Describe the bug

The current bulk delete doesn't clear or flush which is normally used before or after a bulk operation. 

https://github.com/keycloak/keycloak/blob/be28e866b9bdfac39cae13d06a010994a1e058a7/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java#L149-L162

With the JPA map store implementation a clear would detach all entities in the session, what is not intended here. Without the clear, Hibernate will only flush changes to the DB (as it sees that the modified entity is in a table that is being touched by the query), but will still keep the entity earmarked for an optimistic lock update at the end of the transaction to ensure that it hasn't been modified.

### Version

999-SNAPSHOT

### Expected behavior

After a bulk delete, the entities that are in the persistence context and that have been deleted by the bulk operation should also be marked as deleted in the persistence context.

### Actual behavior

Entities are still in the persistence context and might trigger a optimistic lock exception on commit.

### How to Reproduce?

Seen in #12241

### PR DESCRIPTION
**Pull request by ahus1**
This allows removing them from the persistence context on bulk delete.

Closes #12384

### CODE REVIEW

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java [hunk 1]
```diff
@@ -152,14 +176,32 @@ public long delete(QueryParameters<M> queryParameters) {
 
         CriteriaBuilder cb = em.getCriteriaBuilder();
 
+        // Remove all entities that are in the persistence context and that match the criteria.
+        // This avoids calling flush and clear which would detach all other unrelated entities as well.
+        int[] removed = {0};
+        MapModelCriteriaBuilder<String, E, M> mapMcb = queryParameters.getModelCriteriaBuilder().flashToModelCriteriaBuilder(createCriteriaBuilderMap());
+        cacheWithinSession.entrySet().removeIf(entry -> {
+            if (mapMcb.getKeyFilter().test(entry.getKey()) && mapMcb.getEntityFilter().test(entry.getValue())) {
+                em.remove(em.getReference(entityType, UUIDKey.INSTANCE.fromString(entry.getKey())));
```
Comment (vramik) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java:185]: idea: does it make sense to "just" detach the entities (or evict()) which are about to be deleted and let the bulk delete handle the actual deletion? 

That way we'd mark entities in persistence context AND avoid having sql query per entity in the same time.

Comment (sguilhen) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java:185]: I like the idea of activating the bulk delete in the future and keep the suggested change as is. Overall this whole thing gets a +1 from me.

Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java:185]: I've been giving it a thought that those instances that are created through the constructor of the entities that don't live a managed life within the entity manager could be skipped for the bulk delete iteration, still it is difficult for me to reason if they should be cleared from the cache or not; eventually they might "promote" their delegate to a managed entity ... for now I'd like to keep it as is and see if we hit a performance issue (what I don't think we will). 

So I'll put it on today's agenda, and hopefully get an approval either before or after the meeting to get it merged.

Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapKeycloakTransaction.java:185]: If it's possible to activate bulk delete by hibernate's option (I was not aware of it and I was not able to find the option yet), then I agree with the approach.

One additional note: with this change we are modifying behavior for `read(String key)` method as prior this change json field was always loaded as it was expected that it'll be most likely needed. After this change we load whatever is defined within `JpaMapKeycloakTransaction.selectCbConstruct(CriteriaBuilder, Root<RE>)`.

**Review by vramik — APPROVED** [review]

**Review by abstractj — APPROVED** [review]

**Pull request merged**